flink-test-utils-parent/flink-migration-test-utils/src/main/java/org/apache/flink/test/migration/SnapshotGeneratorUtils.java

```diff
@@ -0,0 +1,178 @@
+    }
+
+    private static void executeParameterizedGenerateMethods(
+            Class<?> migrationTestClass,
+            Method method,
+            Method parametersMethod,
+            FlinkVersion version)
+            throws Throwable {
+        method.setAccessible(true);
+        parametersMethod.setAccessible(true);
+        List<TestRule> classRules = getRuleFields(migrationTestClass, ClassRule.class, null);
+
+        executeWithRules(
+                classRules,
+                new Statement() {
+                    @Override
+                    public void evaluate() throws Throwable {
+                        Object migrationTest = createMigrationTest(migrationTestClass);
+                        List<TestRule> rules =
+                                getRuleFields(migrationTestClass, Rule.class, migrationTest);
+                        Collection<?> arguments =
+                                (Collection<?>) parametersMethod.invoke(migrationTest, version);
+                        for (Object argument : arguments) {
+                            executeWithRules(
+                                    rules,
+                                    new Statement() {
+                                        @Override
+                                        public void evaluate() throws Throwable {
+                                            method.invoke(migrationTest, argument);
+                                        }
+                                    });
+                        }
+                    }
+                });
+    }
+
+    private static void executeWithRules(List<TestRule> rules, Statement statement)
+            throws Throwable {
+        new RunRules(statement, rules, Description.EMPTY).evaluate();
+    }
+
+    private static List<TestRule> getRuleFields(
+            Class<?> migrationTestClass, Class<? extends Annotation> tagClass, Object migrationTest)
+            throws IllegalAccessException {
+        List<TestRule> rules = new ArrayList<>();
+
+        Field[] fields = migrationTestClass.getFields();
+        for (Field field : fields) {
+            if (ExternalResource.class.isAssignableFrom(field.getType())
+                    && field.isAnnotationPresent(tagClass)) {
+                field.setAccessible(true);
+                rules.add((ExternalResource) field.get(migrationTest));
+            }
+        }
+
+        return rules;
+    }
+
+    private static Object createMigrationTest(Class<?> migrationTestClass) throws Exception {
+        Constructor<?>[] constructors = migrationTestClass.getDeclaredConstructors();
+        for (Constructor<?> constructor : constructors) {
+            if (constructor.getParameterCount() == 0) {
+                constructor.setAccessible(true);
+                return constructor.newInstance();
+            }
+        }
+
+        // This class does not have a constructor without argument.
+        Constructor<?> constructor = constructors[0];
```

Comment:
It seems test class could have only one constructor, if there are multiple constructors, it will throw the following exception: 

```
java.lang.IllegalArgumentException: Test class can only have one constructor

	at org.junit.runners.model.TestClass.<init>(TestClass.java:48)
	at org.junit.runners.Parameterized$RunnersFactory.<init>(Parameterized.java:370)
	at org.junit.runners.Parameterized$RunnersFactory.<init>(Parameterized.java:360)
	at org.junit.runners.Parameterized.<init>(Parameterized.java:303)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.junit.internal.builders.AnnotatedBuilder.buildRunner(AnnotatedBuilder.java:104)
	at org.junit.internal.builders.AnnotatedBuilder.runnerForClass(AnnotatedBuilder.java:86)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:70)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:37)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:70)
	at org.junit.internal.requests.ClassRequest.createRunner(ClassRequest.java:28)
	at org.junit.internal.requests.MemoizingRequest.getRunner(MemoizingRequest.java:19)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:50)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
```